File: form/form-kernel-go/quotient_test.go
// quotient_test.go — mirrors the assertions in TS quotient.test.ts and
// Rust quotient.rs#[cfg(test)]. Cross-kernel identity: same inputs into
// any kernel's quotient arm produce the same equivalence answers.

package main

import "testing"

// placeholderCarrier — a distinct NodeID to stand in as the carrier
// slot of a QUOTIENT recipe. The carrier's content doesn't drive
// canonicalization in any of the built-in equivalences; it's a
// structural placeholder.
func placeholderCarrier(k *Kernel) NodeID {
	category := NodeID{
		Pkg:   1,
		Level: LevelBasic,
		Type:  RBasicList,
		Inst:  0,
	}
	return k.intern(category, nil)
}

func TestRBasicQuotientIsSlot70(t *testing.T) {
	if RBasicQuotient != 70 {
		t.Fatalf("RBasicQuotient = %d, want 70", RBasicQuotient)
	}
	if RBasicEquivalence != 71 {
		t.Fatalf("RBasicEquivalence = %d, want 71", RBasicEquivalence)
	}
}

func TestIntegerFromNatPairSharesNodeID(t *testing.T) {
	k := NewKernel()
	lib := BuildQuotientLibrary(k)
	carrier := placeholderCarrier(k)
	q := MakeQuotientRecipe(k, carrier, lib.EquivIntegerFromNatPair.NodeID)

	v31 := InternQuotientValue(k, q, []NodeID{k.internTrivialInt(3), k.internTrivialInt(1)})
	v53 := InternQuotientValue(k, q, []NodeID{k.internTrivialInt(5), k.internTrivialInt(3)})
	v97 := InternQuotientValue(k, q, []NodeID{k.internTrivialInt(9), k.internTrivialInt(7)})

	if v31 != v53 {
		t.Errorf("(3,1) ≡ (5,3) [both +2]: %v != %v", v31, v53)
	}
	if v31 != v97 {
		t.Errorf("(3,1) ≡ (9,7) [transitivity]: %v != %v", v31, v97)
	}

	// Negative integers
	vn13 := InternQuotientValue(k, q, []NodeID{k.internTrivialInt(1), k.internTrivialInt(3)})
	vn24 := InternQuotientValue(k, q, []NodeID{k.internTrivialInt(2), k.internTrivialInt(4)})
	if vn13 != vn24 {
		t.Errorf("(1,3) ≡ (2,4) [both -2]: %v != %v", vn13, vn24)
	}
	if v31 == vn13 {
		t.Errorf("+2 == -2 should not collide: %v", v31)
	}
	if !QuotientEqual(k, v31, v53) {
		t.Errorf("QuotientEqual((3,1),(5,3)) should hold")
	}
}

func TestRationalFromIntPairCanonicalizes(t *testing.T) {
	k := NewKernel()
	lib := BuildQuotientLibrary(k)
	carrier := placeholderCarrier(k)
	q := MakeQuotientRecipe(k, carrier, lib.EquivRationalFromIntPair.NodeID)

	v24 := InternQuotientValue(k, q, []NodeID{k.internTrivialInt(2), k.internTrivialInt(4)})
	v12 := InternQuotientValue(k, q, []NodeID{k.internTrivialInt(1), k.internTrivialInt(2)})
	v36 := InternQuotientValue(k, q, []NodeID{k.internTrivialInt(3), k.internTrivialInt(6)})
	vNeg2Pos4 := InternQuotientValue(k, q, []NodeID{k.internTrivialInt(-2), k.internTrivialInt(4)})
	vPos2Neg4 := InternQuotientValue(k, q, []NodeID{k.internTrivialInt(2), k.internTrivialInt(-4)})

	if v24 != v12 {
		t.Errorf("2/4 ≡ 1/2: %v != %v", v24, v12)
	}
	if v36 != v12 {
		t.Errorf("3/6 ≡ 1/2 (reduce): %v != %v", v36, v12)
	}
	if vNeg2Pos4 != vPos2Neg4 {
		t.Errorf("-2/4 ≡ 2/-4 (sign normalization): %v != %v", vNeg2Pos4, vPos2Neg4)
	}
	if v12 == vNeg2Pos4 {
		t.Errorf("1/2 ≠ -1/2 but they share a NodeID: %v", v12)
	}
}

func TestCommutativePairSwaps(t *testing.T) {
	k := NewKernel()
	lib := BuildQuotientLibrary(k)
	carrier := placeholderCarrier(k)
	q := MakeQuotientRecipe(k, carrier, lib.EquivCommutativePair.NodeID)

	a := k.internTrivialInt(7)
	b := k.internTrivialInt(42)
	c := k.internTrivialInt(99)

	vab := InternQuotientValue(k, q, []NodeID{a, b})
	vba := InternQuotientValue(k, q, []NodeID{b, a})
	if vab != vba {
		t.Errorf("(7,42) ≡ (42,7): %v != %v", vab, vba)
	}
	vac := InternQuotientValue(k, q, []NodeID{a, c})
	if vab == vac {
		t.Errorf("(7,42) != (7,99) but they collide: %v", vab)
	}
}

func TestCanonicalRoundTripShape(t *testing.T) {
	k := NewKernel()
	lib := BuildQuotientLibrary(k)
	carrier := placeholderCarrier(k)
	q := MakeQuotientRecipe(k, carrier, lib.EquivIntegerFromNatPair.NodeID)

	v := InternQuotientValue(k, q, []NodeID{k.internTrivialInt(7), k.internTrivialInt(2)})
	canon := CanonicalForm(k, v)
	kids := k.children(canon)
	if len(kids) != 3 {
		t.Fatalf("[quotient, canon-a, canon-b]: got %d children", len(kids))
	}
	ca, ok := readIntTrivial(kids[1])
	if !ok {
		t.Fatalf("canon-a not int trivial")
	}
	cb, ok := readIntTrivial(kids[2])
	if !ok {
		t.Fatalf("canon-b not int trivial")
	}
	if ca != 5 {
		t.Errorf("canon-a = %d, want 5", ca)
	}
	if cb != 0 {
		t.Errorf("canon-b = %d, want 0", cb)
	}
	// Re-intern from canonical children lands at same NodeID
	v2 := InternQuotientValue(k, q, []NodeID{kids[1], kids[2]})
	if v != v2 {
		t.Errorf("canonical re-intern not idempotent: %v != %v", v, v2)
	}
}

func TestEquivalenceCellsAreContentAddressed(t *testing.T) {
	k := NewKernel()
	a := BuildQuotientLibrary(k)
	b := BuildQuotientLibrary(k)
	if a.EquivIntegerFromNatPair.NodeID != b.EquivIntegerFromNatPair.NodeID {
		t.Errorf("same kernel + bootstrap should yield identical equivalence-cell NodeID")
	}
	resolved, ok := ResolveEquivalence(k, a.EquivIntegerFromNatPair.NodeID)
	if !ok {
		t.Fatalf("ResolveEquivalence failed for known cell")
	}
	if resolved.EquivalenceName != "integer-from-nat-pair" {
		t.Errorf("name = %q, want integer-from-nat-pair", resolved.EquivalenceName)
	}
	if resolved.Decidability != uint32(DecidableCheap) {
		t.Errorf("decidability = %d, want %d", resolved.Decidability, DecidableCheap)
	}
}

func TestQuotientRecipesAreContentAddressed(t *testing.T) {
	k := NewKernel()
	lib := BuildQuotientLibrary(k)
	carrier := placeholderCarrier(k)
	q1 := MakeQuotientRecipe(k, carrier, lib.EquivIntegerFromNatPair.NodeID)
	q2 := MakeQuotientRecipe(k, carrier, lib.EquivIntegerFromNatPair.NodeID)
	if q1 != q2 {
		t.Errorf("same (carrier, equiv) should yield same QUOTIENT NodeID: %v != %v", q1, q2)
	}
}

func TestDecidabilityPolicyRoutesStrategy(t *testing.T) {
	k := NewKernel()
	RegisterHandler("test-heavy-g", func(_ *Kernel, raw []NodeID) []NodeID {
		out := make([]NodeID, len(raw))
		copy(out, raw)
		return out
	})
	RegisterHandler("test-undec-g", func(_ *Kernel, raw []NodeID) []NodeID {
		out := make([]NodeID, len(raw))
		copy(out, raw)
		return out
	})

	heavy := MakeEquivalence(k, "test-heavy-g", DecidableHeavy, "test-heavy-g")
	undec := MakeEquivalence(k, "test-undec-g", Undecidable, "test-undec-g")

	if heavy.Strategy != uint32(StrategyLazy) {
		t.Errorf("heavy strategy = %d, want Lazy (%d)", heavy.Strategy, StrategyLazy)
	}
	if undec.Strategy != uint32(StrategyLazy) {
		t.Errorf("undec strategy = %d, want Lazy (%d)", undec.Strategy, StrategyLazy)
	}
	if undec.IsDecidable {
		t.Errorf("undec.IsDecidable should be false")
	}
}

func TestLazyStrategyMergesOnDemand(t *testing.T) {
	k := NewKernel()
	RegisterHandler("lazy-int-g", func(kk *Kernel, raw []NodeID) []NodeID {
		av, ok := readIntTrivial(raw[0])
		if !ok {
			panic("lazy-int-g: bad child 0")
		}
		bv, ok := readIntTrivial(raw[1])
		if !ok {
			panic("lazy-int-g: bad child 1")
		}
		return []NodeID{kk.internTrivialInt(av - bv), kk.internTrivialInt(0)}
	})
	lazyEq := MakeEquivalence(k, "lazy-int-g", DecidableHeavy, "lazy-int-g")
	carrier := placeholderCarrier(k)
	q := MakeQuotientRecipe(k, carrier, lazyEq.NodeID)

	v31 := InternQuotientValue(k, q, []NodeID{k.internTrivialInt(3), k.internTrivialInt(1)})
	v53 := InternQuotientValue(k, q, []NodeID{k.internTrivialInt(5), k.internTrivialInt(3)})

	// Lazy: raw NodeIDs differ pre-canonicalization
	if v31 == v53 {
		t.Errorf("lazy raw NodeIDs should differ pre-canonicalization: %v", v31)
	}
	c31 := CanonicalForm(k, v31)
	c53 := CanonicalForm(k, v53)
	if c31 != c53 {
		t.Errorf("CanonicalForm should merge: %v != %v", c31, c53)
	}
	if !QuotientEqual(k, v31, v53) {
		t.Errorf("QuotientEqual should hold across strategies")
	}
}

func TestQuotientPartsInspection(t *testing.T) {
	k := NewKernel()
	lib := BuildQuotientLibrary(k)
	carrier := placeholderCarrier(k)
	q := MakeQuotientRecipe(k, carrier, lib.EquivCommutativePair.NodeID)

	c, e, err := QuotientParts(k, q)
	if err != nil {
		t.Fatalf("QuotientParts: %v", err)
	}
	if c != carrier {
		t.Errorf("carrier mismatch: %v != %v", c, carrier)
	}
	if e != lib.EquivCommutativePair.NodeID {
		t.Errorf("equiv mismatch: %v != %v", e, lib.EquivCommutativePair.NodeID)
	}
}

func TestHandlerRegistryIsQueryable(t *testing.T) {
	k := NewKernel()
	_ = BuildQuotientLibrary(k)
	if !HasHandler("integer-from-nat-pair") {
		t.Errorf("HasHandler(integer-from-nat-pair) should be true")
	}
	if HasHandler("does-not-exist-handler-name-g") {
		t.Errorf("unknown handler should not register as present")
	}
	fn, ok := GetHandler("integer-from-nat-pair")
	if !ok || fn == nil {
		t.Errorf("GetHandler should return the registered fn")
	}
}

File: form/form-kernel-ts/src/quotient.test.ts
// quotient.test.ts — QUOTIENT arm proof-of-shape tests.
//
// Verifies the core promise: two representatives equivalent under a
// quotient's relation receive the SAME NodeID. Run with:
//
//   npx tsx src/quotient.test.ts
//
// Lightweight assertion harness so we don't pull in Jest/Vitest for the
// quotient conformance folder. Exits non-zero on failure; prints PASS/FAIL lines.

import { Kernel, Level, RBasic, nodeKey, type NodeID } from "./kernel.ts";
import {
  buildQuotientLibrary,
  canonical_form,
  Decidability,
  getHandler,
  intern_quotient_value,
  make_quotient_recipe,
  makeEquivalence,
  quotient_equal,
  quotient_parts,
  registerHandler,
  resolve_equivalence,
  CanonStrategy,
} from "./quotient.ts";

let failures = 0;
let passes = 0;

function ok(name: string, cond: boolean, detail = ""): void {
  if (cond) {
    passes++;
    console.log(`PASS  ${name}`);
  } else {
    failures++;
    console.log(`FAIL  ${name}${detail ? ` — ${detail}` : ""}`);
  }
}

function eqNode(a: NodeID, b: NodeID): boolean {
  return (
    a.pkg === b.pkg &&
    a.level === b.level &&
    a.type === b.type &&
    a.inst === b.inst
  );
}

// ---------------------------------------------------------------------------
// Test 1 — RBasic.QUOTIENT constant value (no collision with INDUCTIVE)
// ---------------------------------------------------------------------------

ok(
  "RBasic.QUOTIENT is slot 70",
  (RBasic as Record<string, number>).QUOTIENT === 70,
  `actual = ${(RBasic as Record<string, number>).QUOTIENT}`,
);

// ---------------------------------------------------------------------------
// Test 2 — Integer-from-nat-pair: (3,1) and (5,3) share NodeID
// ---------------------------------------------------------------------------

{
  const k = new Kernel();
  const lib = buildQuotientLibrary(k);
  // Carrier: just use a placeholder list-recipe.
  const carrier = k.intern(
    { pkg: 1, level: Level.BASIC, type: RBasic.LIST, inst: 0 },
    [],
  );
  const Q = make_quotient_recipe(
    k,
    carrier,
    lib.EQUIV_INTEGER_FROM_NAT_PAIR.nodeID,
  );

  const v31 = intern_quotient_value(k, Q, [
    k.internTrivialInt(3),
    k.internTrivialInt(1),
  ]);
  const v53 = intern_quotient_value(k, Q, [
    k.internTrivialInt(5),
    k.internTrivialInt(3),
  ]);
  const v97 = intern_quotient_value(k, Q, [
    k.internTrivialInt(9),
    k.internTrivialInt(7),
  ]);

  ok(
    "integer-from-nat-pair: (3,1) ≡ (5,3) [both represent +2]",
    eqNode(v31, v53),
    `${nodeKey(v31)} vs ${nodeKey(v53)}`,
  );
  ok(
    "integer-from-nat-pair: (3,1) ≡ (9,7) [transitivity]",
    eqNode(v31, v97),
  );

  // Negative integer: (1, 3) → diff = -2; (2, 4) → diff = -2 → same NodeID
  const vn13 = intern_quotient_value(k, Q, [
    k.internTrivialInt(1),
    k.internTrivialInt(3),
  ]);
  const vn24 = intern_quotient_value(k, Q, [
    k.internTrivialInt(2),
    k.internTrivialInt(4),
  ]);
  ok(
    "integer-from-nat-pair: (1,3) ≡ (2,4) [both represent -2]",
    eqNode(vn13, vn24),
  );

  // Non-equivalent: +2 vs -2 → DIFFERENT NodeID
  ok(
    "integer-from-nat-pair: +2 ≠ -2 NodeID",
    !eqNode(v31, vn13),
  );

  // quotient_equal helper
  ok(
    "quotient_equal: (3,1) == (5,3) via helper",
    quotient_equal(k, v31, v53),
  );
}

// ---------------------------------------------------------------------------
// Test 3 — Rational-from-int-pair: (2,4) and (1,2) share NodeID
// ---------------------------------------------------------------------------

{
  const k = new Kernel();
  const lib = buildQuotientLibrary(k);
  const carrier = k.intern(
    { pkg: 1, level: Level.BASIC, type: RBasic.LIST, inst: 0 },
    [],
  );
  const Q = make_quotient_recipe(
    k,
    carrier,
    lib.EQUIV_RATIONAL_FROM_INT_PAIR.nodeID,
  );

  const v24 = intern_quotient_value(k, Q, [
    k.internTrivialInt(2),
    k.internTrivialInt(4),
  ]);
  const v12 = intern_quotient_value(k, Q, [
    k.internTrivialInt(1),
    k.internTrivialInt(2),
  ]);
  const v36 = intern_quotient_value(k, Q, [
    k.internTrivialInt(3),
    k.internTrivialInt(6),
  ]);
  const v_neg2_4 = intern_quotient_value(k, Q, [
    k.internTrivialInt(-2),
    k.internTrivialInt(4),
  ]);
  const v_2_neg4 = intern_quotient_value(k, Q, [
    k.internTrivialInt(2),
    k.internTrivialInt(-4),
  ]);

  ok("rational-from-int-pair: 2/4 ≡ 1/2", eqNode(v24, v12));
  ok("rational-from-int-pair: 3/6 ≡ 1/2 [reduce]", eqNode(v36, v12));
  ok(
    "rational-from-int-pair: -2/4 ≡ 2/-4 [sign normalization]",
    eqNode(v_neg2_4, v_2_neg4),
  );
  ok(
    "rational-from-int-pair: 1/2 ≠ -1/2 NodeID",
    !eqNode(v12, v_neg2_4),
  );
}

// ---------------------------------------------------------------------------
// Test 4 — Commutative-pair: (a,b) ≡ (b,a)
// ---------------------------------------------------------------------------

{
  const k = new Kernel();
  const lib = buildQuotientLibrary(k);
  const carrier = k.intern(
    { pkg: 1, level: Level.BASIC, type: RBasic.LIST, inst: 0 },
    [],
  );
  const Q = make_quotient_recipe(
    k,
    carrier,
    lib.EQUIV_COMMUTATIVE_PAIR.nodeID,
  );

  const aRaw = k.internTrivialInt(7);
  const bRaw = k.internTrivialInt(42);

  const vab = intern_quotient_value(k, Q, [aRaw, bRaw]);
  const vba = intern_quotient_value(k, Q, [bRaw, aRaw]);

  ok("commutative-pair: (7,42) ≡ (42,7)", eqNode(vab, vba));

  // Different pairs stay distinct.
  const cRaw = k.internTrivialInt(99);
  const vac = intern_quotient_value(k, Q, [aRaw, cRaw]);
  ok("commutative-pair: (7,42) ≠ (7,99) NodeID", !eqNode(vab, vac));
}

// ---------------------------------------------------------------------------
// Test 5 — Round-trip: canonical form decodes back to valid representative
// ---------------------------------------------------------------------------

{
  const k = new Kernel();
  const lib = buildQuotientLibrary(k);
  const carrier = k.intern(
    { pkg: 1, level: Level.BASIC, type: RBasic.LIST, inst: 0 },
    [],
  );
  const Q = make_quotient_recipe(
    k,
    carrier,
    lib.EQUIV_INTEGER_FROM_NAT_PAIR.nodeID,
  );

  const v = intern_quotient_value(k, Q, [
    k.internTrivialInt(7),
    k.internTrivialInt(2),
  ]);
  const canonical = canonical_form(k, v);
  // canonical's children: [quotient_recipe, canonical-a, canonical-b]
  const kids = k.children(canonical);
  ok(
    "round-trip: canonical form has [quotient, canon-a, canon-b] shape",
    kids.length === 3,
  );
  const ca = k.trivialValue(kids[1]!);
  const cb = k.trivialValue(kids[2]!);
  ok(
    "round-trip: canonical-a == diff (7-2=5)",
    ca.kind === "int" && ca.int === 5,
  );
  ok(
    "round-trip: canonical-b == 0",
    cb.kind === "int" && cb.int === 0,
  );

  // Re-intern from canonical form lands at same NodeID
  const v2 = intern_quotient_value(k, Q, [kids[1]!, kids[2]!]);
  ok("round-trip: canonical re-intern is idempotent", eqNode(v, v2));
}

// ---------------------------------------------------------------------------
// Test 6 — Substrate-cell content-addressing for equivalence-recipes
// ---------------------------------------------------------------------------

{
  const k = new Kernel();
  const a = buildQuotientLibrary(k);
  const b = buildQuotientLibrary(k);
  // Same kernel, same bootstrap → same NodeIDs across calls.
  ok(
    "equivalence cells are content-addressed across calls",
    eqNode(
      a.EQUIV_INTEGER_FROM_NAT_PAIR.nodeID,
      b.EQUIV_INTEGER_FROM_NAT_PAIR.nodeID,
    ),
  );

  // Resolve back from cell NodeID
  const resolved = resolve_equivalence(
    k,
    a.EQUIV_INTEGER_FROM_NAT_PAIR.nodeID,
  );
  ok(
    "resolve_equivalence: round-trips equivalence name",
    resolved.equivalence_name === "integer-from-nat-pair",
  );
  ok(
    "resolve_equivalence: round-trips decidability",
    resolved.decidability === Decidability.DECIDABLE_CHEAP,
  );
}

// ---------------------------------------------------------------------------
// Test 7 — Decidability policy: cheap → EAGER, heavy → LAZY
// ---------------------------------------------------------------------------

{
  const k = new Kernel();
  registerHandler(
    "test-heavy",
    (_k, raw) => raw, // identity for the test
  );
  registerHandler(
    "test-undecidable",
    (_k, raw) => raw,
  );

  const heavy = makeEquivalence(k, {
    equivalence_name: "test-heavy",
    decidability: Decidability.DECIDABLE_HEAVY,
    handler_name: "test-heavy",
  });
  const undec = makeEquivalence(k, {
    equivalence_name: "test-undecidable",
    decidability: Decidability.UNDECIDABLE,
    handler_name: "test-undecidable",
  });

  ok(
    "policy: DECIDABLE_HEAVY → LAZY strategy",
    heavy.strategy === CanonStrategy.LAZY,
  );
  ok(
    "policy: UNDECIDABLE → LAZY strategy",
    undec.strategy === CanonStrategy.LAZY,
  );
  ok(
    "policy: UNDECIDABLE → is_decidable=false",
    !undec.is_decidable,
  );
}

// ---------------------------------------------------------------------------
// Test 8 — Lazy strategy still produces equal canonical forms
// ---------------------------------------------------------------------------

{
  const k = new Kernel();
  registerHandler(
    "lazy-integer-pair",
    (kk, raw) => {
      // Same logic as the eager handler — just registered as heavy.
      const av = kk.trivialValue(raw[0]!);
      const bv = kk.trivialValue(raw[1]!);
      if (av.kind !== "int" || bv.kind !== "int") {
        throw new Error("bad children");
      }
      return [kk.internTrivialInt(av.int - bv.int), kk.internTrivialInt(0)];
    },
  );
  const lazyEq = makeEquivalence(k, {
    equivalence_name: "lazy-integer-pair",
    decidability: Decidability.DECIDABLE_HEAVY,
    handler_name: "lazy-integer-pair",
  });
  const carrier = k.intern(
    { pkg: 1, level: Level.BASIC, type: RBasic.LIST, inst: 0 },
    [],
  );
  const Q = make_quotient_recipe(k, carrier, lazyEq.nodeID);

  const v31 = intern_quotient_value(k, Q, [
    k.internTrivialInt(3),
    k.internTrivialInt(1),
  ]);
  const v53 = intern_quotient_value(k, Q, [
    k.internTrivialInt(5),
    k.internTrivialInt(3),
  ]);

  // Lazy: raw NodeIDs differ (different inst=3 entries with different children)
  ok("lazy: raw NodeIDs differ pre-canonicalization", !eqNode(v31, v53));
  // But canonical-form merges them
  ok(
    "lazy: canonical_form merges (3,1) and (5,3)",
    eqNode(canonical_form(k, v31), canonical_form(k, v53)),
  );
  ok(
    "lazy: quotient_equal works regardless of strategy",
    quotient_equal(k, v31, v53),
  );
}

// ---------------------------------------------------------------------------
// Test 9 — quotient_parts inspection
// ---------------------------------------------------------------------------

{
  const k = new Kernel();
  const lib = buildQuotientLibrary(k);
  const carrier = k.intern(
    { pkg: 1, level: Level.BASIC, type: RBasic.LIST, inst: 0 },
    [],
  );
  const Q = make_quotient_recipe(
    k,
    carrier,
    lib.EQUIV_COMMUTATIVE_PAIR.nodeID,
  );

  const parts = quotient_parts(k, Q);
  ok("quotient_parts: extracts carrier", eqNode(parts.carrier, carrier));
  ok(
    "quotient_parts: extracts equivalence",
    eqNode(parts.equivalence, lib.EQUIV_COMMUTATIVE_PAIR.nodeID),
  );
}

// ---------------------------------------------------------------------------
// Test 10 — Handler registry is queryable
// ---------------------------------------------------------------------------

{
  // Force bootstrap by building a library
  const k = new Kernel();
  buildQuotientLibrary(k);
  ok(
    "registry: integer-from-nat-pair handler registered",
    getHandler("integer-from-nat-pair") !== undefined,
  );
  ok(
    "registry: unknown handler returns undefined",
    getHandler("does-not-exist") === undefined,
  );
}

// ---------------------------------------------------------------------------
// Summary
// ---------------------------------------------------------------------------

console.log("");
console.log(`Results: ${passes} passed, ${failures} failed`);
if (failures > 0) process.exit(1);
